src/errorMiddleware/errorMiddleware.ts

```diff
@@ -0,0 +1,65 @@
+import { Context, Middleware } from 'koa';
+
+/**
+ * @see {@link https://github.com/microsoft/TypeScript/issues/1863}
+ */
+const ERROR_STATE_KEY = (Symbol('seek-koala-error') as unknown) as string;
+
+const isObject = (value: unknown): value is Record<PropertyKey, unknown> =>
+  typeof value === 'object' && value !== null;
+
+/**
+ * Catches errors thrown from downstream middleware, as specified here:
+ *
+ * https://github.com/koajs/koa/wiki/Error-Handling#catching-downstream-errors
+ *
+ * This tries to extract a numeric error `status` to serve as the response
+ * status, and will set the error message as the response body for non-5xx
+ * statuses. It works well with Koa's built-in `ctx.throw`.
+ *
+ * All caught errors are emitted to the `error` event:
+ *
+ * ```javascript
+ * app.on('error', (err, ctx) => {});
+ * ```
+ *
+ * This should be placed high up the middleware chain so that errors from lower
+ * middleware are handled. It also serves to set the correct `ctx.status` for
+ * middleware that emit logs or metrics containing the response status. For
+ * this reason, we recommend a sequence like:
+ *
+ * ```javascript
+ * app
+ *   .use(requestLoggingMiddleware)
+ *   .use(metricsMiddleware)
+ *   .use(ErrorMiddleware.handle)
+ *   .use(allTheRest);
+ * ```
+ */
+export const handle: Middleware = async (ctx, next) => {
+  try {
+    return (await next()) as unknown;
+  } catch (err) {
+    ctx.state[ERROR_STATE_KEY] = err as unknown;
+
+    ctx.app.emit('error', err, ctx);
```

Comment:
Should this only be happening with `5xx`? I think otherwise we'll get a lot of error logs for e.g. `ctx.throw(404)`

Reply:
What do you think about dropping the event emitter? I was just following the Koa recommendation, but we don't really have a compelling reason to use it currently when we can use `thrown` in RequestLoggingMiddleware.

Reply:
:+1: I'm happy to just skip emitting the event